src/java/org/apache/cassandra/journal/Flusher.java

```diff
@@ -181,11 +186,11 @@ private void doFlush()
         }
 
         // flush the segment, schedule write callbacks if requested, return whether the segment has been flushed fully
-        private boolean doFlush(ActiveSegment<K> segment, boolean scheduleCallbacks)
+        private boolean doFlush(ActiveSegment<K, V> segment, boolean scheduleCallbacks)
         {
             int syncedOffset = segment.flush();
             if (scheduleCallbacks)
-                segment.scheduleOnSuccessCallbacks(syncedOffset);
+                callbacks.onFlush(segment.descriptor.timestamp, syncedOffset);
             return segment.isFullyFlushed(syncedOffset);
```

Comment:
Agreed. Renamed the method to `isCompletedAndFullyFlushed()` and changed the logic to break out early if an older segment isn't fully complete and flushed. Also to invoke `onFlushed()` callback just once per iteration of the flusher instead of once per segment that'd been flushed.